### PR TITLE
add generic heap implementation

### PR DESCRIPTION
**Pull request by fabioibanez**
PR to add a generic heap data structure that fits into fixed buffers.

Tried to make the tests comparable to data structures implemented with `NodeAllocator`.

### CODE REVIEW

File: src/heap.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        } else if left_childidx <= self.size as usize {
+            if self.nodes[left_childidx].v < self.nodes[rootidx].v {
+                Self::swap_node(&mut self.nodes, rootidx, left_childidx);
+                self.heapify_down_min(left_childidx)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
```
Comment (jarry-xiao) [inline src/heap.rs:193]: Can you move this to the test directory? as well as add some bench tests

File: src/heap.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    pub fn _pop(&mut self) {
+        let lastidx = (self.size - 1) as usize;
+        Self::swap_node(&mut self.nodes, 0, lastidx);
+        self.nodes[(self.size - 1) as usize] = Node::default();
+        self.size -= 1;
+        self._heapifydown(0);
+    }
+}
+
+trait Min<T: PartialOrd + Copy + Clone + Default + Pod + Zeroable> {
```
Comment (jarry-xiao) [inline src/heap.rs:125]: I don't really see why this needs to be a trait

File: src/heap.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
+/*
+General implementation of a heap
+*/
+use bytemuck::{Pod, Zeroable};
+use std::cmp::PartialOrd;
+
+#[derive(Debug, Default, Clone, Copy)]
+pub struct Node<T> {
+    pub v: T,
```
Comment (jarry-xiao) [inline src/heap.rs:9]: What is the purpose of making this a struct? If anything I think you would want to make `Node<K, V>` so you can use `K` as a sorting key. If it's purely `T`, then you definitely don't need `Node`

File: src/heap.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
+/*
+General implementation of a heap
```
Comment (jarry-xiao) [inline src/heap.rs:2]: https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html
If the interface matches this, that would be ideal

File: src/heap.rs [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        }
+        if index == 0 {
+            return;
+        }
+        let index: usize = index;
+        let parent_index = (index - 1) / 2;
+
+        if self.nodes[index].v > self.nodes[parent_index].v {
+            Self::swap_node(&mut self.nodes, index, parent_index);
+            self._heapifyup(parent_index)
```
Comment (jarry-xiao) [inline src/heap.rs:77]: I would recommend implementing this iteratively as opposed to recursively

File: src/heap.rs [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        } else {
+            return;
+        }
+    }
+
+    fn _heapifydown(&mut self, rootidx: usize) {
+        let rootidx = rootidx;
```
Comment (jarry-xiao) [inline src/heap.rs:84]: ?

File: src/lib.rs [hunk 7]
```diff
@@ -4,6 +4,7 @@ pub mod deque;
 pub mod hash_table;
 pub mod node_allocator;
 pub mod red_black_tree;
+pub mod heap;
```
Comment (jarry-xiao) [inline src/lib.rs:7]: `binary_heap` or `priority_queue` would both be better names

File: src/heap.rs [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        let left_childidx = (2 * rootidx) + 1;
+        let right_childidx = (2 * rootidx) + 2;
+
+        if right_childidx <= self.size as usize {
+            if self.nodes[left_childidx].v > self.nodes[right_childidx].v {
+                if self.nodes[left_childidx].v > self.nodes[rootidx].v {
+                    Self::swap_node(&mut self.nodes, rootidx, left_childidx);
+                    self._heapifydown(left_childidx)
```
Comment (jarry-xiao) [inline src/heap.rs:92]: Iterative also preferred here

File: src/heap.rs (hunk 4 above)
Comment (jarry-xiao) [reply src/heap.rs:2]: https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.from

Can you check if this works natively in Solana? I don't see too much of a reason for this data structure in particular to be difficult to read in from an arbitrary buffer

File: src/binary_heap.rs [hunk 9]
```diff
@@ -0,0 +1,389 @@
+use bytemuck::{Pod, Zeroable};
+use core::ops::{Deref, DerefMut, Drop};
+use std::cmp::PartialOrd;
+
+#[derive(Debug, Default, Clone, PartialOrd, PartialEq)]
+pub struct Node<K, V> {
+    pub key: K,
+    pub value: V,
```
Comment (fabioibanez) [inline src/binary_heap.rs:8]: Fixed node struct to include V for arbitrary metadata

File: src/binary_heap.rs [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+                    continue;
+                }
+                break;
+            } else {
+                return;
+            }
+        }
+    }
+
+    fn _heapifydown(&mut self, index: usize) {
```
Comment (fabioibanez) [inline src/binary_heap.rs:131]: iterative implementation of all previously recursive functions ✅

File: src/binary_heap.rs [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    fn deref_mut(&mut self) -> &mut K {
+        debug_assert!(!self.heap._is_empty());
+        self.sift = true;
+        // SAFE: PeekMut is only instantiated for non-empty heaps
+        unsafe { &mut self.heap.nodes.get_unchecked_mut(0).key }
+    }
+}
+
+#[derive(Debug)]
+pub struct BinaryHeapIterator<
```
Comment (fabioibanez) [inline src/binary_heap.rs:362]: Implemented iterator

**Comment by jarry-xiao** [comment]
Please remove the .DS_Store file from the PR

**Comment by jarry-xiao** [comment]
Why are there so many changes in the other files?

File: benches/bench.rs [hunk 12]
```diff
@@ -100,6 +113,19 @@ mod bench_tests {
...
+        let mut sokoban_heap = Heap::<u128, u128, 1001>::default();
+        let mut slice: Vec<u128> = (0..1000).collect();
+
+        b.iter(|| {
+            for v in 0..1000 {
+                sokoban_heap._push(rng.gen::<u128>());
+            }
+        })
+    }
+
```
Comment (jarry-xiao) [inline benches/bench.rs:128]: Can you paste the benchmark stats?

File: src/binary_heap.rs [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+        const MAX_SIZE: usize,
+    > Heap<K, V, MAX_SIZE>
+{
+    fn swap_node(arr: &mut [Node<K, V>; MAX_SIZE], parent_idx: usize, added_idx: usize) {
+        let temp = arr[parent_idx];
+        arr[parent_idx] = arr[added_idx];
+        arr[added_idx] = temp;
+    }
+
+    pub fn _is_empty(&self) -> bool {
```
Comment (jarry-xiao) [inline src/binary_heap.rs:92]: public functions should not start with an `_`

File: src/binary_heap.rs [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+    pub heap: Heap<K, V, MAX_SIZE>,
+    pub current: u64,
+}
+impl<
+        K: PartialOrd + Copy + Clone + Default + Pod + Zeroable,
+        V: Copy + Clone + Default + Pod + Zeroable,
+        const MAX_SIZE: usize,
+    > Iterator for BinaryHeapIterator<K, V, MAX_SIZE>
+{
+    type Item = (K, V);
```
Comment (jarry-xiao) [inline src/binary_heap.rs:376]: These should be references

File: src/binary_heap.rs [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+    }
+}
+
+#[derive(Debug)]
+pub struct BinaryHeapIterator<
+    K: PartialOrd + Copy + Clone + Default + Pod + Zeroable,
+    V: Copy + Clone + Default + Pod + Zeroable,
+    const MAX_SIZE: usize,
+> {
+    pub heap: Heap<K, V, MAX_SIZE>,
```
Comment (jarry-xiao) [inline src/binary_heap.rs:367]: This looks like you're copying the fully heap every time to make an iterator

File: src/binary_heap.rs [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+        let node = Node::<K, V> {
+            key: value,
+            value: V::default(),
+        };
+        self.nodes[self.size as usize] = node;
+        self._heapify_up_min(self.size as usize);
+        self.size += 1;
+    }
+
+    pub fn _pop_min(&mut self) {
```
Comment (jarry-xiao) [inline src/binary_heap.rs:208]: I don't think this index is guaranteed to be the "min"

File: tests/test.rs [hunk 17]
```diff
@@ -160,7 +179,164 @@ async fn test_simulate_avl_tree() {
...
+
+        for i in 0..(MAX_SIZE - 1) {
+            sokoban_heap_arr[i] = sokoban_heap.nodes[i].key;
+        }
+
+        println!("STL Heap");
+        println!("{:?}", stl_heap_to_vector_insert);
+
+        println!("SOKOBAN Heap");
+        println!("{:?}", sokoban_heap_arr);
```
Comment (jarry-xiao) [inline tests/test.rs:269]: please remove these debug print statements from the test file

File: benches/bench.rs (hunk 12 above)
Comment (fabioibanez) [reply benches/bench.rs:128]: test bench_tests::bench_sokoban_binary_heap_insert_1000_u128          ... bench:      44,853 ns/iter (+/- 617)
test bench_tests::bench_sokoban_binary_heap_insert_20000_u128         ... bench:      45,293 ns/iter (+/- 1,492)
test bench_tests::bench_sokoban_binary_heap_remove_1000_u128          ... bench:      91,893 ns/iter (+/- 415)

**Pull request closed without merging**